server/asset/eth/tokener.go

```diff
@@ -0,0 +1,150 @@
+// This code is available on the terms of the project LICENSE.md file,
+// also available online at https://blueoakcouncil.org/license/1.0.0.
+
+//go:build lgpl
+// +build lgpl
+
+package eth
+
+import (
+	"context"
+	"fmt"
+	"math/big"
+
+	"decred.org/dcrdex/dex"
+	"decred.org/dcrdex/dex/networks/erc20"
+	erc20v0 "decred.org/dcrdex/dex/networks/erc20/contracts/v0"
+	dexeth "decred.org/dcrdex/dex/networks/eth"
+	swapv0 "decred.org/dcrdex/dex/networks/eth/contracts/v0"
+	"github.com/ethereum/go-ethereum/accounts/abi/bind"
+	"github.com/ethereum/go-ethereum/common"
+)
+
+// swapContract is a generic source of swap contract data.
+type swapContract interface {
+	Swap(context.Context, [32]byte) (*dexeth.SwapState, error)
+}
+
+// tokenAddresser exposes the TokenAddress method of a token swap contract.
+type tokenAddresser interface {
+	TokenAddress(*bind.CallOpts) (common.Address, error)
+}
+
+// erc2Contract exposes methods of a token's ERC20 contract.
+type erc20Contract interface {
+	BalanceOf(*bind.CallOpts, common.Address) (*big.Int, error)
+}
+
+// tokener is a contract data manager for a token.
+type tokener struct {
+	*registeredToken
+	swapContract
+	tokenAddresser
+	erc20Contract
+	ver                     uint32
+	contractAddr, tokenAddr common.Address
+}
+
+// newTokener is a constructor for a tokener.
+func newTokener(ctx context.Context, assetID uint32, net dex.Network, be bind.ContractBackend) (*tokener, error) {
+	token, netToken, swapContract, err := networkToken(assetID, net)
+	if err != nil {
+		return nil, err
+	}
+
+	if token.ver != 0 {
+		return nil, fmt.Errorf("only version 0 contracts supported")
+	}
+
+	es, err := erc20v0.NewERC20Swap(swapContract.Address, be)
+	if err != nil {
+		return nil, err
+	}
+
+	erc20, err := erc20.NewIERC20(netToken.Address, be)
+	if err != nil {
+		return nil, err
+	}
+
+	tkn := &tokener{
+		ver:             0,
+		registeredToken: token,
+		swapContract:    &swapSourceV0{es},
+		tokenAddresser:  es,
+		erc20Contract:   erc20,
+		contractAddr:    swapContract.Address,
+		tokenAddr:       netToken.Address,
+	}
+
+	if boundAddr, err := tkn.tokenAddress(ctx); err != nil {
+		return nil, fmt.Errorf("error retrieving bound address for %s version %d contract: %w",
+			token.Name, token.ver, err)
+	} else if boundAddr != netToken.Address {
```

Comment:
:+1: OK that jives. But this means we don't need one of either `tokener.tokenAddr` or `tokener.tokenAddresser` I think

Reply:
For my understanding:

We are just checking that the swap contract is for the correct token here correct? Is there a need to save the token's contract address beyond that? We will coordinate with contract address? If same contract must be same token correct, because the token address is hard coded?